### PR TITLE
Added Coffeelint config

### PR DESCRIPTION
**Pull request by yozlet**
Did this to bring CodeClimate's Coffeescript linting in line with our existing Rubocop policy. @afeld wants to discuss various config aspects.

### CODE REVIEW

File: coffeelint.json [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+  "no_throwing_strings": {
+    "level": "error"
+  },
+  "no_trailing_semicolons": {
+    "level": "error"
+  },
+  "no_trailing_whitespace": {
+    "level": "error",
+    "allowed_in_comments": false,
+    "allowed_in_empty_lines": true
```
Comment (jessieay) [inline coffeelint.json]: we want to allow trailing whitespace in empty lines?

File: coffeelint.json [hunk 2]
```diff
@@ -0,0 +1,135 @@
+{
+  "arrow_spacing": {
+    "level": "ignore"
+  },
+  "braces_spacing": {
+    "level": "ignore",
```
Comment (jessieay) [inline coffeelint.json]: I would love if it we did not ignore this!

File: coffeelint.json [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    "level": "error"
+  },
+  "empty_constructor_needs_parens": {
+    "level": "ignore"
+  },
+  "ensure_comprehensions": {
+    "level": "warn"
+  },
+  "eol_last": {
+    "level": "ignore"
```
Comment (jessieay) [inline coffeelint.json]: I would also love to not ignore this!

File: coffeelint.json [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+  },
+  "newlines_after_classes": {
+    "value": 3,
+    "level": "ignore"
+  },
+  "no_backticks": {
+    "level": "error"
+  },
+  "no_debugger": {
+    "level": "warn",
```
Comment (jessieay) [inline coffeelint.json]: what about `error`? isn't it ideal to make sure debugger statements don't make it into prod?

File: coffeelint.json [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    "console": false
+  },
+  "no_empty_functions": {
+    "level": "ignore"
```
Comment (jessieay) [inline coffeelint.json]: seems worthwhile to `error` on this one

> Disallows declaring empty functions. The goal of this rule is that unintentional empty callbacks can be detected:

- http://www.coffeelint.org/

**Comment by yozlet** [comment]
Thumbs up to all @jessieay's suggestions. To clarify: This first commit is literally the default generated config with one change (ignoring `max_line_length`), as opposed to something that reflects my own preferences. So have at it.

File: coffeelint.json [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+  "colon_assignment_spacing": {
+    "level": "ignore",
+    "spacing": {
+      "left": 0,
+      "right": 0
+    }
+  },
+  "cyclomatic_complexity": {
+    "value": 10,
+    "level": "ignore"
```
Comment (afeld) [inline coffeelint.json]: I don't think we want to disable this.

**Comment by afeld** [comment]
Some ideas:

* [ ] Run CoffeeLint as part of CI.
* [ ] Slim the `coffeelint.json` to the non-default rules. It can apparently generate this using `coffeelint --trimconfig`.
* [ ] Mention the linting in [the setup docs](https://github.com/18F/C2/blob/master/doc/setup.md#running-tests).

File: coffeelint.json [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+  },
+  "no_unnecessary_fat_arrows": {
+    "level": "warn"
+  },
+  "non_empty_constructor_needs_parens": {
+    "level": "ignore"
+  },
+  "prefer_english_operator": {
+    "level": "ignore",
```
Comment (afeld) [inline coffeelint.json]: On the fence about this one...definitely good for ensuring that `==` isn't used accidentally, but I don't necessarily prefer `is` to `===`. Thoughts?

File: coffeelint.json [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    "value": "unix"
+  },
+  "max_line_length": {
+    "value": 80,
+    "level": "ignore",
+    "limitComments": true
+  },
+  "missing_fat_arrows": {
+    "level": "ignore",
```
Comment (afeld) [inline coffeelint.json]: On the fence about this one...definitely good for ensuring that instance methods won't accidentally be called in the wrong scope, but it's good practice to take care of ensuring the correct context on the caller side. Thoughts?

**Comment by pkarman** [comment]
I will happily defer to group consensus on all JS formatting guidelines.

**Comment by pkarman** [comment]
@yozlet, @afeld has had changes merged into this PR. If it still reflects what you wanted, I can merge. Please comment here either way.

**Comment by pkarman** [comment]
@yozlet also, @afeld made some suggestions in https://github.com/18F/C2/pull/695#issuecomment-148375257 that probably need addressing, either here or in a separate PR (maybe open a Chore story in PT?).

**Comment by yozlet** [comment]
@pkarman @afeld Yep, this all looks good to me. I confess that I only started this because I got irritated with CodeClimate complaining about one line that was a couple of chars too long, and figured that, hey, if we're enforcing standards for Ruby...

So I *think* that CC will automatically use this config now that it exists, but I may be wrong - I don't know enough about how CC works. If I'm right, that takes care of @afeld's first checkbox, and he's done the second, which only leaves the setup docs.

**Comment by pkarman** [comment]
According to this we might need to explicitly specify our coffeelint config file.
https://github.com/codeclimate/codeclimate-coffeelint/issues/3#issuecomment-123863081

However, I don't see a way to confirm that CC is actually running that engine. Maybe we should introduce an intentional violation to confirm?

**Comment by afeld** [comment]
Should be running:

https://github.com/18F/C2/blob/0933ce24ef2b21c5b595acd94daec45e474831b1/.codeclimate.yml#L10-L13

It is interesting that you can't see the output log from CodeClimate, though...

**Pull request merged**